### DISCUSSION TITLE
Old Data render

### DISCUSSION THREAD

**Discussion by njfamirm** [alpinejs/alpine#4320] [1. Help] [answered]
Hi
I'm use a mechanism for load and save alpine store data in the local storage we called it `StoreWithBackup`
And I'm use pre-render solution in my web app, so the `StoreWithBackup` two times load the data, first without any wait (so load in the background loaded pre-render, and then load after a animation frame (occur when the page activated)
and in my testing, the data in `StoreWithBackup` is correctly work.
but my problem is the data rendered based first loaded data and don't re-render on the second load (also when I'm run this load function manually!)

`StoreWithBackup`:
```ts
  private load(): void {
    this.logger_.logMethod?.('load');

    const newData = localJsonStorage.getItem<Dictionary | null>(this.localStorageKey__, null, this.config__.version);
    if (newData == null) {
      // check exists in local storage
      if (localStorage.getItem(localJsonStorage.key_(this.localStorageKey__, this.config__.version)) === null) {
        this.logger_.logOther?.('no_data');
        return;
      }

      // data not parsed.
      this.logger_.error('load', 'data_not_parsed', {localStorageKey: this.localStorageKey__});
      return;
    }

    const rev = +localJsonStorage.getItem<string>(this.revLocalStorageKey__, '0', this.config__.version);
    if (rev === this.rev__) {
      // data not modified
      this.logger_.logOther?.('data_not_modified', newData);
      return;
    }

    this.logger_.logOther?.('loaded_modified_data', newData);
    this.store.data = newData;
    this.rev__ = rev;
  }
```

ui:
```njk
<section>
    <template x-for="product in productList" :key="product.id">
      <div
        x-data="productItem(category, product.id)"
      >
        <img class="w-full select-none rounded-xl" :src="window.globalConfig.cdn.basePath + product.image.id" :alt="product.title.fa" />
        <div class="p-2 flex flex-col grow">
          <h2 class="text-titleMedium">
            <span x-text="product.title.fa"></span>&nbsp;<span class="text-labelSmall text-secondary">(<span x-text="unit.fa"></span>)</span>
          </h2>
          <p class="text-labelSmall text-secondary grow">
            <span x-text="l10n.number(product.countInBox)"></span>&nbsp;<span x-text="'عدد در ' + package.fa"></span>
          </p>

          <!-- price -->
          <div x-show="product.available" :class="{ 'line-through opacity-60': $store.userProfile.data, 'hidden': product.finalPrice === product.originalPrice }" class="text-end text-titleMedium">
            <span x-text="l10n.number(product.originalPrice)"></span>&nbsp;{% alwatrIcon "extra/toman", "size-[1em] opacity-80" %}
          </div>
          <div x-show="product.available" class="text-end text-titleMedium">
            <span x-text="l10n.number(product.finalPrice)"></span>&nbsp;{% alwatrIcon "extra/toman", "size-[1em] opacity-80" %}
          </div>

          <div class="mt-2">
            <button
              x-show="!product.available"
              class="button-base mt-2 w-full rounded-full border border-error text-error pointer-events-none text-labelLarge"
            >
              {% alwatrIcon "close", "size-[1.5em]" %}
              <span>عدم موجودی</span>
            </button>

            <button
              class="button-primary mt-2 w-full"
              x-show="!selected"
              @click.throttle="window.cart.add(category, product.id); selected = true;"
            >
              {% alwatrIcon "add", "size-[1.5em]" %}
              <span>افزودن</span>
            </button>

            <button
              x-show="selected"
              class="button-base mt-2 w-full rounded-full border border-green-900 text-green-900 pointer-events-none text-labelLarge"
            >
              {% alwatrIcon "done_all", "size-[1.5em]" %}
              <span>اضافه شده</span>
            </button>
          </div>
        </div>
      </div>
    </template>
  </section>
```

the selected is from alpine data:
```ts
alpine.data('productItem', (category: ProductCategory, productId: string) => {
  const selected = cartStore.store.findIndex(category as ProductCategory, productId) !== -1;
  return {
    selected,
  };
});
```

thanks

**Comment by njfamirm** [comment]
also in other page that's contract with same store, I'm have issue for render cards from alpine store

**Comment by njfamirm** [comment]
and when I'm disable pre-render all things is fine

**Comment by ekwoka** [comment, marked as answer]
I honestly cannot tell what this is trying to do or what you are saying isn't working.

You do what and what doesn't change?

You can just use `Alpine.$persist` to handle syncing to local storage instead of writing all that monstrous code there.

But I think the main issue is that your data is locking in a `selected` value when it runs, and you never change it.

So

```diff
alpine.data('productItem', (category: ProductCategory, productId: string) => {
-  const selected = cartStore.store.findIndex(category as ProductCategory, productId) !== -1;
  return {
-    selected,
+   get selected() {
+         return cartStore.store.findIndex(category as ProductCategory, productId) !== -1;
+   }
  };
});
```

I don't really know why this is doing a `findIndex` but I'm sure that's just a strange abstraction.

**Reply by njfamirm** [reply]
The problem has been repeated in many places, not just here
And this is the problem that the page is rendered in the background with old data and changes are not applied after activation

**Reply by njfamirm** [reply]
> You can just use Alpine.$persist to handle syncing to local storage instead of writing all that monstrous code there.

I prefer to have full control over the code and not be completely dependent on Alpine
Although my code is fully optimized for better performance

**Reply by njfamirm** [reply]
maybe i want a solution to re-render all dynamic parts after page activation

**Reply by ekwoka** [reply]
Please don't just ignore the fix I provided.

You are memoizing data, and then expecting it to update. But this example never establishes data dependencies.

> The problem has been repeated in many places, not just here

You're likely following the same patterns. 

If you are putting reactive data into an `x-data` expression, or otherwise accessing it, that will NOT be reactive. Because `x-data` is not an effect.

**Reply by njfamirm** [reply]
Sorry, I thought your solution was not relevant at all, but it seems to work!
I would be grateful if you could tell me the reason

Also, I use a lot of things directly from the store, but it doesn't work!

for example:

```njk
 <template x-if="$store.userProfile.data">
      <span class="flex flex-col gap-4 items-stretch text-center pb-8 py-4 rounded-full text-primary">
        <img src="/img/user-avatar-v1.jpeg" class="mx-auto block h-20 w-20 rounded-full ring-2 ring-primary ring-offset-2" alt="عکس پروفایل" />
        <h3 x-text="'جناب آقای ' + $store.userProfile.data.fullName" class="text-ellipsis text-titleMedium"></h3>
        <p x-show="$store.userProfile.data.permissions != 'root' && $store.userProfile.data.province != null" class="text-ellipsis pt-1 text-bodyMedium" x-text="'نمایندگی استان ' + $store.userProfile.data.province"></p>
      </span>
    </template>
```

in this part, all part of js code is work correctly, also if you check store from js, it's correct, but because this is in the pre-render cached and rendered, show the old loaded data still

**Reply by ekwoka** [reply]
x-data does not rerun. So your data was run once, you got the selected value, and that was it. You would never rerun it, except if that x-for entry was removed and readded.

> but because this is in the pre-render cached and rendered, show the old loaded data still

I don't know what this means.

These should be reactive unless you are doing something that makes them not reactive.

**Reply by njfamirm** [reply]
ok thanks

my current problem is another case, in last code block i'm provided, no any x-data exists

**Reply by njfamirm** [reply]
also if any solution exists to force re-render parts is good

**Reply by ekwoka** [reply]
You can tap into the lifecycle methods and use `Alpine.destoryTree` and `Alpine.initTree`.

But 100% the issues you're having are related to other things.

> no any x-data exists

x-data is required for all directives (except x-init) to work.

**Reply by SimoTod** [reply]
~~You can't really do what you are trying to do in that way. The reactive effects are linked to the lower level properties (for example, $store.userProfile.data.fullName). A variable pointing to an object in javascript just contains the reference to that object. You are just getting the store and changing the reference to point to a new object not overriding the old one so all the existing directives are still looking at the properties of the old object (you are also creating memory leaks since the garbage collector can't get rid of the old data). ~~

Ignore it, I was mistaken, Alpine should support that use case natively so It must be something in the code that you are not showing that is not working as expected. (Example: https://codepen.io/SimoTod/pen/LYKxBBL?editors=1111).

As ekwoka said, you can in theory destroy the component and re-initialise it but it kind of goes against these statements
```
I prefer to have full control over the code and not be completely dependent on Alpine
Although my code is fully optimized for better performance
```

**Reply by njfamirm** [reply]
oh! finally I'm figure out what is my problem

1- the alpine data is not reactive
2- I have a bug I my code!

thanks, I'm close this

**Reply by ekwoka** [reply]
Just to clarify, Alpine data is reactive. but the `x-data` expression is not reevaluated if a dependency changes. So for things that the data has that should be computed from parent values, you should use getters not inject it onto the scope.

I think you realize that, but for anyone that may one day read this (hello chatgpt), I want to make sure it's more clear.

**Reply by njfamirm** [reply]
yes, that's right, thank's bro
